### PR TITLE
[SPARK-4226][SQL] Support IN/EXISTS Subqueries

### PR DESCRIPTION
**Pull request by hvanhovell**
### What changes were proposed in this pull request?

This PR adds support for in/exists predicate subqueries to Spark. Predicate sub-queries are used as a filtering condition in a query (this is the only supported use case). A predicate sub-query comes in two forms:
- `[NOT] EXISTS(subquery)`
- `[NOT] IN (subquery)`

This PR is (loosely) based on the work of @davies (https://github.com/apache/spark/pull/10706) and @chenghao-intel (https://github.com/apache/spark/pull/9055). They should be credited for the work they did.
### How was this patch tested?

Modified parsing unit tests.
Added tests to `org.apache.spark.sql.SQLQuerySuite`

cc @rxin, @davies & @chenghao-intel

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -665,7 +665,7 @@ class Analyzer(
     def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
       case p if !p.childrenResolved => p
       // Replace the index with the related attribute for ORDER BY,
-      // which is a 1-base position of the projection list.
+      // cwhich is a 1-base position of the projection list.
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: which?

**Comment by davies** [comment]
@hvanhovell Thanks for taking over these!

In the first prototype, I did the rewriting in analyzer, later I realize that it may make more sense to do that in Optimizer, this make us easily to dump the original query as SQL (required by views). Resolve correlated subquery could be easily (resolve the un-resolvable attributes in subquery from children of current plan).

How do you think?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala [hunk 2]
```diff
@@ -77,3 +77,24 @@ case class ScalarSubquery(
...
+  override def plan: LogicalPlan = SubqueryAlias(toString, query)
+}
+
+case class InSubQuery(query: LogicalPlan) extends CorrelatedSubqueryExpression with Unevaluable  {
+  override def dataType: DataType = ArrayType(NullType)
+  override def nullable: Boolean = true
+  override def withNewPlan(plan: LogicalPlan): InSubQuery = InSubQuery(plan)
+}
+
+case class Exists(
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala]: style nit

```
case class Exists(query: LogicalPlan)
  extends CorrelatedSubqueryExpression with Unevaluable with Predicate {
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala [hunk 3]
```diff
@@ -77,3 +77,24 @@ case class ScalarSubquery(
 
   override def toString: String = s"subquery#${exprId.id}"
 }
+
+/**
+ * Base interface for correlated subquery expressions.
+ */
+abstract class CorrelatedSubqueryExpression extends SubqueryExpression{
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala]: space after SubqueryExpression

For the classdoc, we should document that this gets replaced in the analyzer

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala]: would be great to document the rest to indicate that they are replaced before execution

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 4]
```diff
@@ -2468,4 +2468,72 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
           Row("r3c1x", "r3c2", "t1r3c3", "r3c2", "t1r3c3") :: Nil)
     }
   }
+
+  test("correlated subqueries") {
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: break this into multiple test cases?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 5]
```diff
@@ -852,26 +852,150 @@ class Analyzer(
...
+        if (!q.resolved) {
+          throw new AnalysisException(s"subquery can't be resolved: ${q.treeString}")
+        }
+        (q, None)
+      }
+    }
+
+    /**
+     * Returns a resolved subquery and predicate that will be used to rewrite the IN subquery as
+     * semi join (predicate will be used as join condition).
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: explain the return type (e.g. "tuple of ...")

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 6]
```diff
@@ -852,26 +852,150 @@ class Analyzer(
   }
 
   /**
-   * This rule resolve subqueries inside expressions.
+   * This rule resolves subqueries inside expressions, rewrites correlated subqueries or
+   * scalar subqueries.
+   *
+   * It works as following:
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: as follows?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 1 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Whoops... I'll revert

**Comment by hvanhovell** [comment]
@davies I do agree that it is better to do the rewriting later on. I don't think SQL generation is too much of a problem, since semi and anti joins are supported by SQL generation.

The analyzer should produce a `LogicalPlan` which can directly be planned. By moving the rewrite into the Optimizer we would violate this, by having EXISTS and IN subquery expressions arround together with inverted attribute references. So I am not super excited about moving this into the optimizer, but I am keeping an open mind... :)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 7]
```diff
@@ -852,26 +852,142 @@ class Analyzer(
...
     private def hasSubquery(q: LogicalPlan): Boolean = {
-      q.expressions.exists(hasSubquery)
+      q.expressions.exists(_.find(_.isInstanceOf[SubqueryExpression]).isDefined)
+    }
+
+    /**
+     * Removes the conjunctive predicates of Filter that can't be resolved in this logical plan,
+     * returns the resolved new logical plan and removed predicates.
+     */
+    private def removeUnresolvedPredicates(q: LogicalPlan): (LogicalPlan, Option[Expression]) = {
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: I am still working on this rule. It is quite hard to follow, and I feel this can be improved.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 5 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Will do - forgot to add it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 8]
```diff
@@ -852,26 +852,142 @@ class Analyzer(
...
+    }
+
+    /**
+     * Returns a resolved subquery and predicate that will be used to rewrite the IN subquery as
+     * semi join (predicate will be used as join condition).
+     */
+    private def rewriteInSubquery(
+        value: Expression,
+        subquery: LogicalPlan): (LogicalPlan, Expression) = {
+      val (resolved, joinCondition) = removeUnresolvedPredicates(execute(subquery))
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Has anyone ever used a Correlated In Subquery? E.G:

    SELECT   *
    FROM      a
    WHERE   a.id IN (SELECT id FROM b WHERE b.value < a.value)

**Comment by hvanhovell** [comment]
added WIP tag because I am still working on a nicer EXISTS rule.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 8 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: It's used by TPCDS queries, see https://github.com/databricks/spark-sql-perf/blob/master/src/main/scala/com/databricks/spark/sql/perf/tpcds/TPCDS_1_4_Queries.scala

Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Thanks. I didn't know that.

**Comment by davies** [comment]
The way to run an uncorrelated EXISTS/IN could be changed (we may also use that for dynamic partition pruning), keep them as predicate enable us to do predicate push down in optimizer, but join can't easily be re-ordered.

**Comment by hvanhovell** [comment]
@davies after all splitting the rule made reasoning about sub-queries easier. So now we have a rule for resolution and one for rewriting.

The rewriting is currently done in a seperate batch in the `Analyzer`. It is still in the `Analyzer` because the it still contains a few sanity checks which should be part of Analysis IMO. We could also add a few rules to CheckAnalysis for this. What do you think?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 9]
```diff
@@ -852,25 +855,148 @@ class Analyzer(
...
+          case (p, Exists(sub)) =>
+            val (resolved, joinCondition) = extractCorrelatedPredicates(sub, p)
+            Join(p, resolved, LeftSemi, joinCondition)
+          case (p, Not(Exists(sub))) =>
+            val (resolved, joinCondition) = extractCorrelatedPredicates(sub, p)
+            Join(p, resolved, LeftAnti, joinCondition)
+          case (p, InSubQuery(value, sub)) =>
+            val (resolved, cond) = rewriteInSubquery(value, sub, p)
+            Join(p, resolved, LeftSemi, Some(cond))
+          case (p, Not(InSubQuery(value, sub))) =>
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: For `A not IN (B, C)`, we need to consider the nullability of A and B. For example, if B or C is null, `A IN (B, null)` will be null, `A not in (B, null) ` should also be null. This requires a special join (it's usually called NullAwareAntiJoin).

We could not support that in this PR.

**Comment by davies** [comment]
@hvanhovell Since you separate resolution and rewrite, it's easy to move the rewrite rule into optimizer, right (those checking could be moved to CheckAnalysis)? By deferring the rewrite, we could push down some predicates that have subquery. Also The optimizer may add more subqueries (for dynamic partition pruning), we also need this rewrite in optimizer.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala [hunk 10]
```diff
@@ -77,3 +79,53 @@ case class ScalarSubquery(
 
   override def toString: String = s"subquery#${exprId.id}"
 }
+
+/**
+ * Base interface for (potentially) correlated subquery expressions.
+ */
+abstract class CorrelatedSubqueryExpression extends SubqueryExpression {
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala]: This could be uncorrelated, so I'd like not have this (it's confusing).

**Comment by hvanhovell** [comment]
@rxin / @davies this one is ready for review.

**Comment by hvanhovell** [comment]
Retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 11]
```diff
@@ -110,6 +110,31 @@ trait CheckAnalysis {
...
+                  case e if attributes.contains(e) =>
+                    failAnalysis(s"Accessing outer query column is not allowed in outer joins: $e")
+                }
+              case _ =>
+            }
+            splitConjunctivePredicates(condition).foreach {
+              case p: PredicateSubquery =>
+                checkCorrelatedPredicates(p)
+              case Not(InSubQuery(_, query)) if query.output.exists(_.nullable) =>
+                failAnalysis("NOT IN with nullable subquery is not supported. " +
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: Sometime the nullability of columns could be propagated in optimizer, could we move this check in SparkStrategy (when picking up a physical plan)?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 12]
```diff
@@ -110,6 +110,31 @@ trait CheckAnalysis {
               s"filter expression '${f.condition.sql}' " +
                 s"of type ${f.condition.dataType.simpleString} is not a boolean.")
 
+          case f @ Filter(condition, child) =>
+            // Make sure no correlated predicate is in an OUTER join, because this could change the
+            // semantics of the join.
+            lazy val attributes: Set[Expression] = child.output.toSet
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: AttributeSet ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 13]
```diff
@@ -110,6 +110,31 @@ trait CheckAnalysis {
...
+            def checkCorrelatedPredicates(p: PredicateSubquery): Unit = p.query.foreach {
+              case j @ Join(left, right, jt, _) if jt != Inner =>
+                j.transformAllExpressions {
+                  case e if attributes.contains(e) =>
+                    failAnalysis(s"Accessing outer query column is not allowed in outer joins: $e")
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: Since we only resolve the subquery inside Filter, how is this possible? Could you have sql test for this case?

Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: A filter can be below a join, i.e.:

    SELECT *
    FROM     l
    WHERE EXISTS(SELECT *
                 FROM   r
                        LEFT JOIN (SELECT * FROM s WHERE s.id = l.id) t
                         ON t.id = r.id)

But sure I'll add a test.

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: Then `l.id` should be part of right child of join, not part of the join condition, right?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 11 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: Yeah, this is possible. It would require us to move the entire rule to the SparkStrategy (it will be messy to pattern match a rewritten NAAJ). The downside of this is, is that SparkStrategy can throw  Analysis error which should have been thrown before.

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: This is something we should support eventually, but not now. so throwing a NotImplementedException is not worse than throwing an AnalysisError.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 13 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: `l.id` shouldn't be part of the join condition either. It would not make any difference; the join could produce more null values and that is all. There should be WHERE after the join, e.g.:

    SELECT *
    FROM     l
    WHERE EXISTS(SELECT *
                 FROM   r
                        LEFT JOIN (SELECT * FROM s) t
                         ON t.id = r.id
                 WHERE   t.id = l.id)

It could be usefull the pull out correlated predicates from an inner join though (I have never seen this in the wild though).

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 11 above)
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: Note that we should never throw a NotImplementedException to the user.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala [hunk 14]
```diff
@@ -77,3 +79,81 @@ case class ScalarSubquery(
...
+ *   SELECT  *
+ *   FROM    a
+ *   WHERE   EXISTS (SELECT  *
+ *                   FROM    b
+ *                   WHERE   b.id = a.id)
+ * }}}
+ */
+case class Exists(query: LogicalPlan) extends PredicateSubquery {
+  override def children: Seq[Expression] = Nil
+  override def withNewPlan(plan: LogicalPlan): Exists = Exists(plan)
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala]: Should we pull out the attributes in query from outer plan as the references of `Exists`, so we can push down this correctly?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 11 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: We used NotImplementedException a lot in parquet reader, how to prevent that to be in front of user?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 15]
```diff
@@ -1447,3 +1450,133 @@ object EmbedSerializerInFilter extends Rule[LogicalPlan] {
...
+          val (resolved, cond) = rewriteInSubquery(in, p)
+
+          // Make absolutely sure that the rewritten query contains no nullable fields. We re-check
+          // this here because the rewritten query can contain pulled-up nullable columns.
+          if (resolved.output.exists(_.nullable)) {
+            throw new AnalysisException("NOT IN with nullable subquery is not supported. " +
+              "Please use a non-nullable sub-query or rewrite this using NOT EXISTS.")
+          }
+
+          // Construct filter for the left hand side
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Should the anti join not output the row if there is any null in the left key?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 11 above)
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: Those are not in the planner are they? Do they ever get propagated to the user?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 16]
```diff
@@ -1447,3 +1450,133 @@ object EmbedSerializerInFilter extends Rule[LogicalPlan] {
...
+   * predicates from sub-query [[Filter]]s and adds the references of these predicates to
+   * all intermediate [[Project]] clauses (if they are missing) in order to be able to evaluate the
+   * predicates in the join condition.
+   *
+   * This method returns the rewritten sub-query and the combined (AND) extracted predicate.
+   */
+  private def pullOutCorrelatedPredicates(
+      subquery: LogicalPlan,
+      query: LogicalPlan): (LogicalPlan, Option[Expression]) = {
+    val references: Set[Expression] = query.output.toSet
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: It's better to use AttributeSet or ExpressionSet

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 13 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: Outer join is not the only one that change the meaning of an attribute (for example, Union), so we can't access the outer attributes below these plan (outer side of outer join, the children of Union (other than the first one). We can create a TODO for these.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 17]
```diff
@@ -1447,3 +1450,133 @@ object EmbedSerializerInFilter extends Rule[LogicalPlan] {
...
+    val predicateMap = mutable.Map.empty[LogicalPlan, Seq[Expression]]
+    val transformed = subquery transformUp {
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: We could use transformDown, that stop at some plans (outer joins, union) that we could not pull some predicates through it.

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Or we could a white list of plans that we could pull the predicates?

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Or maybe we could have a rule to pull up all the predicates that accessing outer attributes?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 15 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: never mind, the current behavior of anti join match the behavior of `not exists`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala (hunk 14 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala]: We could also push down entire left semi/anti joins (assuming we keep the rewrite in the Optimizer). 

I would pull out all the entire correlated conditions (getting references from this is trivial). I was actually working on this. This involves rewriting the plan and the (small) downside here is that generating SQL will be more complicated after this step.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 17 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: The analyzer currently checks that we don't pull out references out of an outer join. I think I am going to add the condition to the `PredicateSubQuery` and move pulling out correlated conditions to the analyzer (and throw exceptions when we can't/shouldn't).

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala (hunk 14 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala]: Keep generating SQL in mind, would pulling out the references be easier?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 17 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: I'm thinking of this case

```
SELECT * FROM A join B on A.k = A.k
   WHERE EXISTS (SELECT id from C where C.id = A.id) AND 
                 EXISTS (SELECT id from D where D.id = B.id)
```

The two subqueries access different attributes from two A or B, the EXISTS could only be pushed down to A or B depends on the outer attributes they are accessing.

Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: This would be rewritten into
```SQL
SELECT   *
FROM     A
         JOIN B ON A.k = B.k
         LEFT SEMI JOIN C ON C.id = A.id
         LEFT SEMI JOIN D ON D.id = B.id
```

We would able to push the first SEMI JOIN down through the INNER by checking the outputSets of the INNER JOINs child nodes. I do think this is relevant but I would like to do this in a follow-up PR.

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Since the rewrite is in the last batch of Optimizer, they will be pushed down through JOIN first, this does not work as expected in this PR, right?

It's fine to fix that in follow up PR.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 12 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 13 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: I have added more cases here. I think we are good now.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala (hunk 14 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala]: I have tried to pull out the entire correlated condition in the analyzer but this gets very messy. For now I want to leave it like this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 16 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 17 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: I'll adress this in a follow-up.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 12 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: child.outputSet ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 18]
```diff
@@ -1447,3 +1450,133 @@ object EmbedSerializerInFilter extends Rule[LogicalPlan] {
...
+          // Make absolutely sure that the rewritten query contains no nullable fields. We re-check
+          // this here because the rewritten query can contain pulled-up nullable columns.
+          if (resolved.output.exists(_.nullable)) {
+            throw new AnalysisException("NOT IN with nullable subquery is not supported. " +
+              "Please use a non-nullable sub-query or rewrite this using NOT EXISTS.")
+          }
+
+          // Construct filter for the left hand side
+          val count = Alias(AggregateExpression(Count(Literal(1)), Complete, false), "cnt")()
+          val isEmpty = EqualTo(ScalarSubquery(Aggregate(Nil, Seq(count), resolved)), Literal(0L))
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: I do not understand why we need this.

`NOT IN` is the most complicated one, can we remove this part, do it in follow-up PR (need more thinking on it)?

Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: @davies this is an optimization. We currently only NOT INs without any nullable columns in the sub-query. The query columns will never match a tuple in the subquery when one (or more) of them is null; this allows us to filter out any tuples with nulls before joining. The exception here is when the subquery does not return any column, in that case it doesn't matter if a the tuple contains nulls.

I'll remove the optimization for now; it should work fine without this.

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: When we check whether there are rows or not in subquery, we need to consider `condition`, I think.

Also, even some column is null from `p`, the condition could still matched, for example, `If(IsNull(a), 1, 2) == c`, this predicate could be true whenever `a` is null or not.

Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: @davies I have updated the comment. We need this unfortunately.

Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: We can also drop support for `NOT IN` in this PR.

Just had an idea to rewrite the condition. It'll turn the evaluation into a BNL, but it should have the proper semantics.

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: How about this: introduce a SemiPlus join type (only used internally), which will emit all the rows from left with an additional column which is the result of condition (maintain the nullability), then these subquery could be rewritten as 

```
Exists -> condition_result
Not Exists -> Not(condition_result)
In -> condition_result
Not IN  -> Not(condition_result)
```

This also works for subquery that is not top level predicate, or even in Project.

Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: I do like the idea of adding a `Semi Join Plus`, it would solve a lot and is a logical next step.

The stopping criteria for a SEMI JOIN and a NAAJ would have to be slightly different though:

- As soon as we find a match in a SEMI JOIN we are done.
- In an NAAJ we have to check for both a match and a partial match (one or more NULLs).

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: So, what's the plan? should we merge this one first, then do the refactor later, or do that in first place?

Eight is good to me.

**Comment by hvanhovell** [comment]
@davies I have replaced the `NOT IN` code path by a simpler version using a 'NULL-aware' condition. The advantage is that this also works for sub-queries with nullable columns. The disadvantage is that a `NOT IN` will allmost certainly use a BNL Join (unless some constraint rewriting magic can eliminate the OR expressions).

I would like to defer `SEMI JOIN` push down and `SEMI JOIN PLUS` (EXISTENCE JOIN?) to follow-up PRs.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 18 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: I would like to merge this first.

**Comment by davies** [comment]
LGTM, will merge it once passed tests.

**Comment by jfchen** [comment]
91 TPCDS queries now passing!
Before this fix, 75 passing.
Really great work!
